### PR TITLE
test: Add kerberized sudo to s4u proof of concept

### PR DESCRIPTION
**Pull request by Gundersanne**
- [x] needs to be released (upstream PR: SSSD/sssd#5367)
- [x] get it on our fedora-33 image (next image rebuild after Feb 12)

### CODE REVIEW

**Comment by martinpitt** [comment]
This was released into Fedora 33 [yesterday](https://bodhi.fedoraproject.org/updates/FEDORA-2021-b4adb66f7d), in sssd 2.4.1. So it's not yet on our fedora-33 image. However, it *is* on the fedora-testing image in https://github.com/cockpit-project/bots/pull/1668 .

@Gundersanne: What is that "/etc/pam.d/teststack"?

**Comment by Gundersanne** [comment]
Nice, I'll try it out see if it works out of the box. I'll remove `teststack`, it's just a pam stack to use with `pamtester` to only test gss.

File: test/verify/check-system-s4u-ssh [hunk 1]
```diff
@@ -90,8 +94,33 @@ gssapi.creds.rcred_cred_store.store_cred_into({ 'ccache'.encode(): 'FILE:/etc/co
...
+        sshd_machine.execute(script="""#!/bin/sh
+        sed -i '/\[domain\/cockpit.lan\]/ a pam_gssapi_services = sudo, sudo-i' /etc/sssd/sssd.conf
+        systemctl restart sssd
+        usermod -G wheel user
+        su user
+        echo -ne 'foobarfoo\nfoobarfoo\nfoobarfoo\n' | kinit -c /home/user/cache user
+        KRB5CCNAME=/home/user/cache sudo echo hai
+        exit 0
+        """)
+        client_machine.execute("KRB5_TRACE=\"/dev/stderr\" KRB5CCNAME=\"/etc/cockpit.ccache\" ssh -vvv -K  user@sshserver.cockpit.lan bash -c 'KRB5CCNAME=/home/user/cache sudo echo hai'")
```
Comment (martinpitt) [inline test/verify/check-system-s4u-ssh:123]: Could you please run `sudo whoami` instead, and check that the returned output contains `root`? That would ensure that it actually worked.

I don't quite understand the difference between the script execution above, and this line? They seem to do similar things?

Comment (Gundersanne) [reply test/verify/check-system-s4u-ssh:123]: I'll do that, the second ssh-ing from the client machine via gssapi auth and then using the cache built above to sudo, putting the 2 things together. But it's indeed just testing the same thing.

**Comment by martinpitt** [comment]
[does not work on rhel-8-4](https://logs.cockpit-project.org/logs/pull-14782-20210222-140653-670d3478-rhel-8-4/log.html#79) and [fedora-33](https://logs.cockpit-project.org/logs/pull-14782-20210222-140653-670d3478-fedora-33/log.html#79). I think it did before, so there is something more subtle -- perhaps more stdout?

I suggest to not just  run `| grep -q`, as this gives zero info in the failure case. rather do something like this:
```py
out = m.execute('...')
self.assertEqual(out, "root")
```

File: test/verify/check-system-s4u-ssh [hunk 2]
```diff
@@ -90,8 +94,35 @@ gssapi.creds.rcred_cred_store.store_cred_into({ 'ccache'.encode(): 'FILE:/etc/co
...
+session    required     pam_limits.so
+session    include      system-auth
+""")
+
+        sshd_machine.execute(script="""#!/bin/sh
+        sed -i '/\[domain\/cockpit.lan\]/ a pam_gssapi_services = sudo, sudo-i' /etc/sssd/sssd.conf
+        systemctl restart sssd
+        usermod -G wheel user
+        su user
+        echo -ne 'foobarfoo\nfoobarfoo\nfoobarfoo\n' | kinit -c /home/user/cache user
```
Comment (martinpitt) [inline test/verify/check-system-s4u-ssh:119]: Why does that need to input the password three times? Forced password change?

File: test/verify/check-system-s4u-ssh [hunk 3]
```diff
@@ -90,8 +94,35 @@ gssapi.creds.rcred_cred_store.store_cred_into({ 'ccache'.encode(): 'FILE:/etc/co
         # enable gssapiauth
         sshd_machine.execute("sed -ri 's/#GSSAPIAuthentication (yes|no)/GSSAPIAuthentication yes/' /etc/ssh/sshd_config")
         sshd_machine.execute("systemctl daemon-reload && systemctl restart sshd")
+        client_machine.execute("KRB5RCACHEDIR=\"/var/cache/krb5rcache\" KRB5_TRACE=\"/dev/stderr\" KRB5CCNAME=\"/etc/cockpit.ccache\" ssh -vvv -K  user@sshserver.cockpit.lan echo hello")
 
-        client_machine.execute("KRB5_TRACE=\"/dev/stderr\" KRB5CCNAME=\"/etc/cockpit.ccache\" ssh -vvv -K  user@sshserver.cockpit.lan echo hello")
+        # configure gssapi authentication for sudo
+        sshd_machine.write('/etc/sudoers.d/cockpit-krb5-sudo', """
+Defaults    env_keep += "KRB5CCNAME"
```
Comment (martinpitt) [inline test/verify/check-system-s4u-ssh:101]: This is rather non-obvious. I wonder how we are going to deal with that for the real implementation.. But that is a discussion for another day.

**Review by martinpitt — APPROVED** [review]
Thank you so much for figuring this out and driving this with the sssd team! Let's get this in to make sure it does not regress, and then we know that the underpinnings are working. We can discuss further simplification of the setup with the sssd team later on.

**Pull request merged**